src/Features/LanguageServer/ProtocolUnitTests/Configuration/DidChangeConfigurationNotificationHandlerTest.cs

```diff
@@ -0,0 +1,240 @@
+using Microsoft.CodeAnalysis.LanguageServer.Handler.Configuration;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Newtonsoft.Json.Linq;
+using Roslyn.Test.Utilities;
+using Roslyn.Utilities;
+using StreamJsonRpc;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.UnitTests.Configuration
+{
+    public class DidChangeConfigurationNotificationHandlerTest : AbstractLanguageServerProtocolTests
+    {
+        // A regex help to check the message we send to client.
+        // It should look like "feature_group.feature_name"
+        private static readonly string s_clientSideSectionPattern = @"^((csharp|visual_basic)\.)?([\w_]+)\.([\w_]+)$";
+
+        public DidChangeConfigurationNotificationHandlerTest(ITestOutputHelper? testOutputHelper) : base(testOutputHelper)
+        {
+        }
+
+        [Fact]
+        public async Task VerifyWorkflow()
+        {
+            var markup = @"
+public class A { }";
+
+            var clientCapabilities = new ClientCapabilities()
+            {
+                Workspace = new WorkspaceClientCapabilities()
+                {
+                    DidChangeConfiguration = new DynamicRegistrationSetting() { DynamicRegistration = true }
+                }
+            };
+
+            var clientCallbackTarget = new ClientCallbackTarget();
+            var initializationOptions = new InitializationOptions()
+            {
+                CallInitialized = true,
+                ClientCapabilities = clientCapabilities,
+                ServerKind = WellKnownLspServerKinds.CSharpVisualBasicLspServer,
+                ClientTarget = clientCallbackTarget,
+            };
+
+            // Let client has non-default values for all options.
+            clientCallbackTarget.SetClientSideOptionValues(setToDefaultValue: false);
+
+            // 1. When initialized, server should register workspace/didChangeConfiguration if client support DynamicRegistration
+            var server = await CreateTestLspServerAsync(
+                markup,
+                initializationOptions: initializationOptions);
+
+            Assert.True(clientCallbackTarget.WorkspaceDidChangeConfigurationRegistered);
+
+            // 2. Server should fetched all the values from client, options should have non-default values now.
+            VerifyValuesInServer(server.TestWorkspace, clientCallbackTarget.MockClientSideValues);
+
+            // 3. When client side value changes, client should send server didChangeConfiguration. Then server would re-fetch all the options.
+            // Let client has a default values for all options.
+            clientCallbackTarget.SetClientSideOptionValues(setToDefaultValue: true);
+
+            await server.ExecuteRequestAsync<DidChangeConfigurationParams, object>(Methods.WorkspaceDidChangeConfigurationName, new DidChangeConfigurationParams(), CancellationToken.None).ConfigureAwait(false);
```

Comment:
since we're not waiting anymore in tests here - does the handler need to even import the operation listener?

Reply:
Ah I forget to remote it : )